controllers/common/flagd-injector.go

```diff
@@ -0,0 +1,459 @@
+		ProviderID: source.ProviderID,
+		Selector:   source.Selector,
+	}
+}
+
+func (fi *FlagdContainerInjector) toFlagdProxyConfig(ctx context.Context, objectMeta *metav1.ObjectMeta, source v1alpha1.Source) (types.SourceConfig, error) {
+	// does the proxy exist
+	exists, ready, err := fi.isFlagdProxyReady(ctx)
+	if err != nil {
+		return types.SourceConfig{}, err
+	}
+	if !exists || (exists && !ready) {
+		return types.SourceConfig{}, constant.ErrFlagdProxyNotReady
+	}
+	ns, n := utils.ParseAnnotation(source.Source, objectMeta.Namespace)
+	return types.SourceConfig{
+		Provider: "grpc",
+		Selector: fmt.Sprintf("core.openfeature.dev/%s/%s", ns, n),
+		URI:      fmt.Sprintf("grpc://%s.%s.svc.cluster.local:%d", FlagdProxyServiceName, fi.FlagdProxyConfig.Namespace, fi.FlagdProxyConfig.Port),
+	}, nil
+}
+
+func (fi *FlagdContainerInjector) isFlagdProxyReady(ctx context.Context) (bool, bool, error) {
+	d := appsV1.Deployment{}
+	err := fi.Client.Get(ctx, client.ObjectKey{Name: FlagdProxyDeploymentName, Namespace: fi.FlagdProxyConfig.Namespace}, &d)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// does not exist, is not ready, no error
+			return false, false, nil
+		}
+		// does not exist, is not ready, is in error
+		return false, false, err
+	}
+	if d.Status.ReadyReplicas == 0 {
+		// exists, not ready, no error
+		if d.CreationTimestamp.Time.Before(time.Now().Add(-3 * time.Minute)) {
+			return true, false, fmt.Errorf(
+				"flagd-proxy not ready after 3 minutes, was created at %s: %w",
+				d.CreationTimestamp.Time.String(),
+				constant.ErrFlagdProxyNotReady,
+			)
+		}
+		return true, false, nil
+	}
+	// exists, at least one replica ready, no error
+	return true, true, nil
+}
+
+func (fi *FlagdContainerInjector) toKubernetesProviderConfig(ctx context.Context, objectMeta *metav1.ObjectMeta, podSpec *corev1.PodSpec, source v1alpha1.Source) (types.SourceConfig, error) {
+	ns, n := utils.ParseAnnotation(source.Source, objectMeta.Namespace)
+
+	// ensure that the FeatureFlagConfiguration exists
+	ff := fi.getFeatureFlag(ctx, ns, n)
+	if ff.Name == "" {
+		return types.SourceConfig{}, fmt.Errorf("feature flag configuration %s/%s not found", ns, n)
+	}
+
+	// add permissions to pod
+	if err := fi.EnableClusterRoleBinding(ctx, objectMeta.Namespace, podSpec.ServiceAccountName); err != nil {
+		return types.SourceConfig{}, err
+	}
+
+	// mark pod with annotation (required to backfill permissions if they are dropped)
+	if objectMeta.Annotations == nil {
+		objectMeta.Annotations = map[string]string{}
+	}
+	objectMeta.Annotations[fmt.Sprintf("%s/%s", constant.OpenFeatureAnnotationPrefix, constant.AllowKubernetesSyncAnnotation)] = "true"
+
+	// build K8s config
+	return types.SourceConfig{
+		URI:      fmt.Sprintf("%s/%s", ns, n),
+		Provider: string(v1alpha1.SyncProviderKubernetes),
+	}, nil
+}
+
+func (fi *FlagdContainerInjector) getFeatureFlag(ctx context.Context, namespace string, name string) v1alpha1.FeatureFlagConfiguration {
```

Comment:
Seems to be a duplicate of common `func FeatureFlag(ctx context.Context, c client.Client, namespace string, name string)`